ipc: intel_adsp: rework host IPC service backend and add critical send support

tmleman opened the pull request:
This pull request extends the generic IPC service API with a function to support sending critical messages/notifications and refactors the Intel Audio DSP host IPC service backend.

Summary of changes:

- ipc_service: Add ipc_service_send_critical() API and a matching send_critical() backend hook to send high-priority messages via a dedicated fast path that bypasses normal busy checks.
- ipc: intel_adsp: Simplify the host IPC service backend by replacing the custom intel_adsp_ipc_msg descriptor and enum-based length encoding with a fixed two-word (header + payload) IPC payload used consistently in send and receive paths.
- intel_adsp: Wire the emergency/critical send path through the new ipc_service_send_critical() API while keeping the normal IPC flow unchanged.
- intel_adsp: Move the synchronous IPC wait logic into the common intel_adsp_ipc_send_message_sync() helper used by tests, and remove the now-redundant backend-specific sync helper.
- intel_adsp: Update and clarify backend documentation comments to describe the new data format, the role of ipc_ept_cfg::priv, and the semantics of the TX-buffer readiness and RX-buffer release helpers.

Status / TODO:

- ~~Existing intel_adsp IPC tests have not yet been adapted to the refactored backend and will currently fail. Test updates will be added as part of this PR before it is ready to merge.~~

tmleman committed:
ipc_service: Add ipc_service_send_critical function

Add ipc_service_send_critical() as a dedicated API for sending critical
high-priority messages over an IPC endpoint.

Introduce an optional send_critical() callback in struct
ipc_service_backend so backends can implement a special fast path that
bypasses normal state and busy checks for critical notifications such as
crash reports or fatal errors.

The ipc_service_send_critical() wrapper mirrors ipc_service_send() on
the service side and delegates the actual behavior to the
backend-specific send_critical() implementation.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman committed:
ipc: intel_adsp: simplify host IPC service backend

Rework the Intel Audio DSP host IPC service backend to use the generic
ipc_service data/len conventions and the new critical send and buffer
management APIs.

Replace the backend-specific intel_adsp_ipc_msg and related enums with a
simple two-word payload passed through the standard ipc_service_send()
and ipc_service_send_critical() interfaces, and adapt the ISR and
receive callback to operate on this representation.

Use ipc_service_get_tx_buffer_size() as a readiness check for the host
channel and provide hold_rx_buffer() / release_rx_buffer()
implementations so ipc_service_release_rx_buffer() can be used to signal
when the channel becomes available again.

Wire the emergency send path through send_critical() to route urgent
messages via the backend's emergency fast path while keeping the normal
IPC flow unchanged.

Move the synchronous IPC wait logic out of the Intel Audio DSP host IPC
backend and into the common intel_adsp_ipc helper used by tests.

Update intel_adsp_ipc_send_message_sync() to both send the IPC message
through ipc_service_send() and wait on the backend semaphore, and remove
the now redundant ipc_send_message_sync() helper from the host IPC
backend.

Document that intel_adsp_ipc_send_message_sync() is a test-only helper,
not used by SOF firmware, and that it is a candidate for future removal
or for replacement by an explicit synchronous send primitive in the
generic IPC service API or in application code.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman committed:
intel_adsp: document host IPC backend data model and send hook

This patch refreshes the documentation in the Intel Audio DSP host IPC
service backend to match the current data model and backend API
contracts.

The updated comments clarify that ipc_ept_cfg::priv must point to
intel_adsp_ipc_ept_priv_data so that the backend can carry state between
the ISR and the application callbacks. They also describe that both the
send and receive paths operate on a fixed two-word uint32_t IPC payload
(header and extended payload) rather than a generic byte buffer, and
that len must always be sizeof(uint32_t) * 2.

The documentation for intel_adsp_ipc_send() is rewritten to explain the
expected payload format, the length and NULL checks performed by the
backend, and how the function programs the IPC registers while relying
on ipc_send_message() for BUSY and TX acknowledgment handling. The
parameter semantics and error returns are aligned with the generic
ipc_service send() hook contract to make the backend behaviour easier to
understand and reuse.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

kv2019i reviewed (commented):
All in all looks good. I think using the service interface rx/tx buffer ops is cleaner mapping. Some questions on whether we keep a struct for the payload (which is still special) and a question on the synchronous send variant.

tmleman commented on soc/intel/intel_adsp/common/ipc.c:
```diff
@@ -43,94 +43,56 @@ static void intel_adsp_ipc_receive_cb(const void *data, size_t len, void *priv)
-		return ret;
-	}
+	ARG_UNUSED(dev);
+	uint32_t msg[2] = {data, ext_data};
```
I haven't seen the changes you're proposing before. I was considering a similar solution, but instead of a structure with a pointer, I saw it more as a buffer that has space for two dwords of message + payload. When SOF sends a message, the IPC backend would know how many bytes to copy to the mailbox based on the passed size.

tmleman committed:
tests: intel_adsp: update smoke IPC tests for new backend

This patch updates the intel_adsp smoke tests to use the new host IPC
backend data model and APIs.

The host IPC smoke test now treats messages as a fixed two-word uint32_t
payload instead of using struct intel_adsp_ipc_msg and the
INTEL_ADSP_IPC_CB_* message types. The receive callback validates the
payload length and pointer, checks the header/payload values against the
expected RETURN_MSG_* constants, and sets msg_flag on receipt. The
test-specific intel_adsp_ipc_send_message() and
intel_adsp_ipc_send_message_sync() helpers are rewritten to send a
two-word payload via ipc_service_send(), and the synchronous helper
waits on the backend semaphore only when the send succeeds. A dedicated
done callback is registered via intel_adsp_ipc_set_done_handler() to
drive done_flag, and intel_adsp_ipc_complete() now completes RX by
calling ipc_service_release_rx_buffer() and asserting on failure.
Completion polling is aligned with the backend by using
ipc_service_get_tx_buffer_size().

The clock calibration smoke test is adapted in the same way. It now
includes the ipc_service and intel_adsp host IPC backend headers when
using the new interface, interprets incoming data as a two-word uint32_t
payload, validates the length and pointer, and stores the returned
timestamp from the extended payload word into the endpoint private data.
The rest of the test flow, including endpoint registration, timestamp
requests and clock rate checks, remains unchanged but runs on top of the
new IPC backend representation.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

kv2019i reviewed (dismissed):
Test results look good, +1 from me.

@tmleman maybe remove the "RFC" tag and "WIP" from summary to gather more feedback.

lyakh commented on soc/intel/intel_adsp/common/ipc.c:
```diff
@@ -43,94 +43,57 @@ static void intel_adsp_ipc_receive_cb(const void *data, size_t len, void *priv)
+	__ASSERT(len == sizeof(uint32_t) * 2, "Unexpected IPC payload length: %zu", len);
+	__ASSERT(data != NULL, "IPC payload pointer is NULL");
+
+	const uint32_t *msg = (const uint32_t *)data;
```
looks like the type-cast shouldn't be needed

lyakh commented on soc/intel/intel_adsp/common/ipc.c:
```diff
@@ -43,94 +43,56 @@ static void intel_adsp_ipc_receive_cb(const void *data, size_t len, void *priv)
+	uint32_t msg[2] = {data, ext_data};
 
-	return 0;
+	return ipc_service_send(&intel_adsp_ipc_ept, &msg, sizeof(msg));
```
in SOF IPC sending will remain synchronous though?

tmleman replied:
@kv2019i We haven't made a decision yet about what we do with 'intel_adsp_ipc_send_message_sync'. I would probably go in the direction of removing this function, I don't think we ever had a use case for it.

lyakh commented on subsys/ipc/ipc_service/backends/ipc_intel_adsp_host_ipc.c:
```diff
@@ -301,51 +288,94 @@ static int ipc_send_message_emergency(const struct device *dev, uint32_t data, u
-		ret = -EBADMSG;
 
-		break;
+	if ((len != (sizeof(uint32_t) * 2)) || data == NULL) {
```
ditto parentheses

softwarecki commented on include/zephyr/ipc/ipc_service_backend.h:
```diff
@@ -71,6 +71,29 @@ struct ipc_service_backend {
+	 *  @param[in] token Backend-specific token.
+	 *  @param[in] data Pointer to the buffer to send.
+	 *  @param[in] len Number of bytes to send.
+	 *
```
Missing -EIO

softwarecki reviewed (approved):
Moving the semaphore handling to `intel_adsp_ipc_send_message_sync` is a good change. It removes the need for `ipc_send_message_sync` and simplifies the flow. Using `hold/release_rx_buffer` for ipc completion is also a clean approach. I like it! I would prefer to keep the `struct intel_adsp_ipc_msg`. The tuple is less readable and any future extension will require updating `sizeof(uint32_t) * 2` in many places.
